synapse/rest/client/v1/room.py

```diff
@@ -266,6 +267,254 @@ def on_PUT(self, request, room_id, event_type, txn_id):
+        # mapping from (type, state_key) -> state_event_id
+        prev_state_map = await self.state_store.get_state_ids_for_event(
+            most_recent_prev_event_id
+        )
+        # List of state event ID's
+        prev_state_ids = list(prev_state_map.values())
+        auth_event_ids = prev_state_ids
+
+        for stateEv in body["state_events_at_start"]:
+            assert_params_in_dict(
+                stateEv, ["type", "origin_server_ts", "content", "sender"]
+            )
+
+            event_dict = {
+                "type": stateEv["type"],
+                "origin_server_ts": stateEv["origin_server_ts"],
+                "content": stateEv["content"],
+                "room_id": room_id,
+                "sender": stateEv["sender"],
+                "state_key": stateEv["state_key"],
+            }
+
+            # Make the state events float off on their own
+            fake_prev_event_id = "$" + random_string(43)
+
+            # TODO: This is pretty much the same as some other code to handle inserting state in this file
+            if event_dict["type"] == EventTypes.Member:
+                membership = event_dict["content"].get("membership", None)
+                event_id, _ = await self.room_member_handler.update_membership(
+                    requester,
+                    target=UserID.from_string(event_dict["state_key"]),
+                    room_id=room_id,
+                    action=membership,
+                    content=event_dict["content"],
+                    outlier=True,
+                    prev_event_ids=[fake_prev_event_id],
+                    # Make sure to use a copy of this list because we modify it
+                    # later in the loop here. Otherwise it will be the same
+                    # reference and also update in the event when we append later.
+                    auth_event_ids=auth_event_ids.copy(),
+                )
+            else:
+                # TODO: Add some complement tests that adds state that is not member joins
+                # and will use this code path. Maybe we only want to support join state events
+                # and can get rid of this `else`?
+                (
+                    event,
+                    _,
+                ) = await self.event_creation_handler.create_and_send_nonmember_event(
+                    requester,
+                    event_dict,
+                    outlier=True,
+                    prev_event_ids=[fake_prev_event_id],
+                    # Make sure to use a copy of this list because we modify it
+                    # later in the loop here. Otherwise it will be the same
+                    # reference and also update in the event when we append later.
+                    auth_event_ids=auth_event_ids.copy(),
+                )
+                event_id = event.event_id
+
+            auth_event_ids.append(event_id)
+
+        logger.info("bulk insert events %s", body["events"])
```

Comment:
We should move this to `DEBUG` or only log the number of events, etc, before landing this.

Reply:
I've removed the various `logger.info` used for my own debugging while working on this.

But I can add some proper `logger.debug` to the interesting areas :+1: